src/contrib/cluster/Akka.Cluster.Tools/PublishSubscribe/Internal/Topics.cs

```diff
@@ -344,40 +346,119 @@ internal static class Utils
         /// </para>
         /// </summary>
         /// <param name="message">TBD</param>
-        /// <returns>TBD</returns>
+        [MethodImpl(MethodImplOptions.NoInlining)]
         public static object WrapIfNeeded(object message)
         {
             return message is RouterEnvelope ? new MediatorRouterEnvelope(message) : message;
         }
+        
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        public static string? KeyToEncodedTopic(string key, string topicPrefix)
+        {
+            if (!key.StartsWith(topicPrefix)) 
+                return null;
+                    
+            var topic = key[(topicPrefix.Length + 1)..];
+            return !topic.Contains('/') ? topic : null;
+        }
+    }
+    
+    /// <summary>
+    /// INTERNAL API
+    ///
+    /// Used for generating Uri-safe topic and group names.
+    /// </summary>
+    internal sealed class PubSubCache
+    {
+        private readonly record struct MakeKeyInfo(ActorPath Path, string Topic);
+        
+        private static readonly System.Text.RegularExpressions.Regex PathRegex = new("^/remote/.+(/user/.+)");
 
-        /// <summary>
-        /// TBD
-        /// </summary>
-        /// <param name="actorRef">TBD</param>
-        /// <returns>TBD</returns>
+        private readonly Dictionary<string, string> _topicToEncodedMap = new();
+        private readonly Dictionary<string, string> _encodedToTopicMap = new();
+        private readonly Dictionary<MakeKeyInfo, string> _makeKeyMap = new();
+        private readonly Dictionary<string, MakeKeyInfo> _makeKeyReverseMap = new();
+        
+        [MethodImpl(MethodImplOptions.NoInlining)]
         public static string MakeKey(IActorRef actorRef)
         {
-            return MakeKey(actorRef.Path);
+            return PathRegex.Replace(actorRef.Path.ToStringWithoutAddress(), "$1");
         }
 
         /// <summary>
         /// TBD
         /// </summary>
         /// <param name="name">TBD</param>
         /// <returns>TBD</returns>
-        public static string EncodeName(string name)
+        public string EncodeName(string name)
         {
-            return name == null ? null : Uri.EscapeDataString(name);
+            if (string.IsNullOrWhiteSpace(name))
+                return null;
+            
+            if (_topicToEncodedMap.TryGetValue(name, out var encoded))
+                return encoded;
+
+            encoded = Uri.EscapeDataString(name);
+            _topicToEncodedMap[name] = encoded;
+            _encodedToTopicMap[encoded] = name;
+            return encoded;
         }
 
-        /// <summary>
-        /// TBD
-        /// </summary>
-        /// <param name="path">TBD</param>
-        /// <returns>TBD</returns>
-        public static string MakeKey(ActorPath path)
+        public string MakeKey(ActorPath path, string topic)
```

Comment:
You could still make this static and do inlining, btw - just refactor these to be extension methods instead.

Reply:
I'm trying to avoid extension methods like a plague since it requires the outermost class declaration to be public static, didn't want to pollute the public API with internal API methods.